File: packs_xtra/logicmoo_packages/prolog/cec/src/cec/prolog/interrupt.pl
/*
 *	file:		interrupt.pl
 *	version:	1.0
 *	date:		March 9, 1990
 *	creation:	March 9, 1990
 *	author:		Uwe Waldmann (uwe)
 *
 *	description:
 *	(see below)
 *
 *	history:
 *	900322	js	Added parameter to establish_handler
 *			Cleaned up inconsistencies in naming
 *			Created module interface
 *
 *	Copyright (C) 1990
 *		Hubert Bertling
 *		Harald Ganzinger
 *		Renate Schaefers
 *		University of Dortmund
 */

% The predicates in this file are used to control the interrupt menu of
% the cec system.  During completion this menu is enhanced by a further
% option "q" ("quit completion").  If the user chooses this option, the
% completion should be terminated as soon as possible.
% 
% 
% Most of the predicates use a global C variable "quit_completion"
% having three posible states:
% 
%    DISABLED      On typing ^C, the user gets the standard interrupt menu.
%
%    ENABLED       On typing ^C, the user has the additional possibility
%                  to type "q" in order to indicate that the completion
%                  should be terminated.
%
%    USED          The "q" option has already been used
% 
% 
% Description of the predicates:
% 
%    establish_handler(+X)
%	installs the new interrupt handler.
%	Set X = 0 for Quintus Prolog Development System
%		  (trace and debug options enabled)
%	or  X = 1 for Quintus Prolog Runtime Generator
%		  (trace and debug options disabled)
%
%    completion_interrupted
%	succeeds, if quit_completion = USED.
%  
%    completion_interrupted(-X)        
%	unifies X with 0, if quit_completion = DISABLED,
%	unifies X with 1, if quit_completion = ENABLED,
%	unifies X with 2, if quit_completion = USED.
%  
%    enable_q_option     quit_completion := ENABLED; succeeds.
%                        (This predicate should be called immediately
%                        after the completion has been started.)
%  
%    disable_q_option    quit_completion := DISABLED; succeeds.
%                        (This predicate should be called immediately
%                        before the completion terminates (no matter whether
%                        it succeeds, fails, or calls "abort"!)% 
%
% 
% The following options from the interrupt menu may modify quit_completion:
% 
%    q   (quit completion)   if quit_completion = ENABLED
%                            then quit_completion := USED.
%  
%    a   (abort)             quit_completion := DISABLED.
% 
% All other options do not change quit_completion.
% 

:- module(interrupt,[establish_handler/1,
                     completion_interrupted/0,
                     disable_q_option/0,
                     enable_q_option/0]).


foreign_file('interrupt.o', [establish_handler,
			     completion_interrupted,
			     disable_q_option,
			     enable_q_option]).

foreign(establish_handler, c, establish_handler(+integer)).
foreign(completion_interrupted, c, completion_interrupted([-integer])).
foreign(disable_q_option, c, disable_q_option).
foreign(enable_q_option, c, enable_q_option).

:- load_foreign_files(['interrupt.o'], []).


completion_interrupted :- completion_interrupted(2).

File: packs_xtra/logicmoo_packages/prolog/cec/src/cec/prolog/interrupt.c
/*
 *	file:		interrupt.c
 *	version:	1.0
 *	date:		March 9, 1990
 *	creation:	March 9, 1990
 *	author:		Uwe Waldmann (uwe)
 *
 *	description:
 *	C functions to be used in connection with interrupt.pl
 *
 *	history:
 *	900322	js	removed absolute pathname for qpaction.h,
 *			use -I option for cc.
 *
 *	Copyright (C) 1990
 *		Hubert Bertling
 *		Harald Ganzinger
 *		Renate Schaefers
 *		University of Dortmund
 */

#include <stdio.h>
#include <signal.h>
#include "quintus.h"

#ifdef QPROLOG_VER_242
#define QP_flush QP_FLUSH
#endif

#define DISABLED 0
#define ENABLED  1
#define USED     2

static long int quit_completion = DISABLED;
static long int rg_version = 0;

/* the following two functions are modified from quintus/library/critical.c */

/*----------------------------------------------------------------------+
|   get_char() reads a (case-folded) character from stdin.		|
+----------------------------------------------------------------------*/

static int get_char()
    {
	char c, command;

	fflush(stdin);
	command = '\n';		/* have not seen a printable character yet */
	read(0, &c, 1);
	while (c != EOF && c != '\n') {
	    if (c > ' ' && c < 127 && command == '\n') {
		command = c;	/* first printable character */
	    }
	    read(0, &c, 1);
	}
	fflush(stdin);
	if (command >= 'A' && command <= 'Z') command += 'a'-'A';
	return command;
    }


static void give_menu()
    {
	for (;;) {
	    if (quit_completion == ENABLED)
	      { printf("\nCompletion interruption (h for help)? "); }
	    else
	      { printf( "\nProlog interruption (h for help)? "); };
	    switch (get_char()) {
		case 'c':
		    /* fflush(stderr) */; return;
		case 'q':
		    if (quit_completion == DISABLED) {
		        printf( "\
\"q\" option disabled\n");
		    } else if (quit_completion == ENABLED) {
		        quit_completion = USED;
			printf("\
\n\n\nThe completion process will be aborted at the next consistent state.\n");
			return;
		    } else {
		        printf( "\
\"q\" option has already been used\n");
		        return;
		    }
		    break;
		case 't':
		    if(!rg_version) QP_action(QP_TRACE);
		    return;
		case 'd':
		    if(!rg_version) QP_action(QP_DEBUG);
		    return;
		case EOF:	/* better solution needed... */
		case 'a':
		    if (quit_completion == ENABLED)
		      { quit_completion = USED;
			printf("\
\n\n\nThe completion process will be aborted at the next consistent state.\n");
			return;
		      }
		    else 
		      { quit_completion = DISABLED;
			QP_action(QP_ABORT);
		      }
		case 'e':
		    QP_action(QP_EXIT);
		case 'h':
		case '?':
		    printf( "\n\
Prolog interrupt options:\n");
		    if (quit_completion == ENABLED) {
		        printf( "\
    q    quit     - quit cec completion\n");
		    }
		    printf( "\
    c    continue - do nothing\n");
		    if( !rg_version) {
			printf( "\
    t    trace    - debugger will start creeping\n\
    d    debug    - debugger will start leaping\n");
		    }
		    printf( "\
    a    abort    - cause a Prolog abort\n\
    e    exit     - irreversible exit from Prolog\n\
    h    help     - this list\n");
		    break;
		default:
		    break;
	    }
	}
    }



void establish_handler(v)
long int v;
{
    rg_version = v;
    signal(SIGINT, give_menu);
}


long int completion_interrupted()
{
    return (quit_completion);
}


void disable_q_option()
{
    quit_completion = DISABLED;
}


void enable_q_option()
{
    quit_completion = ENABLED;
}
